libs/effects-ng/src/lib/directive-effects.spec.ts

```diff
@@ -0,0 +1,99 @@
+import { Component, Injectable } from '@angular/core';
+import { TestBed } from '@angular/core/testing';
+import { createAction, createEffect } from '@ngneat/effects';
+import { tap } from 'rxjs';
+import { ofType } from 'ts-action-operators';
+import { Actions } from './actions';
+import { provideEffectsManager } from './provide-effects-manager';
+import { directiveEffects } from './directive-effects';
+
+const spy = jest.fn();
+const loadTodos = createAction('[Todos] Load Todos');
+
+@Injectable()
+class EffectsOne {
+  loadTodos$ = createEffect((actions) =>
+    actions.pipe(ofType(loadTodos), tap(spy))
+  );
+}
+
+@Injectable()
+class EffectsTwo {}
+
+@Injectable()
+class EffectsThree {}
+
+@Component({
+  template: '',
+  standalone: true,
+  hostDirectives: [
+    directiveEffects(EffectsOne),
+    directiveEffects(EffectsTwo, EffectsThree),
```

Comment:
People should be cautious when using this functionality with multiple component instances (e.g., `ngFor`).

There should be a check to prevent subscribing to the same effects twice, and it would be useful to add a validation when you do this in dev mode.

Reply:
It looks like a global mapper that stores effects class types and a number of their created instances. If the first instance is being created then we subscribe to the effects. If the last one is about to be destroyed then we unsubscribe.

Reply:
I think that you didn't get my point. Let's take the following case:

```html
<app-todos />
<app-todos />
<app-todos />
<app-todos />
```
You'll have multiple subscriptions to the same effect.

Reply:
I got it. I meant that if we create one component multiple times then an effects class will be created multiple times so we can't check effects in the class by reference or via DI(since it's different injectors). To be able to know that we've subscribed to some effects already we need to have some `store` that would say that an effects class had been already instantiated once hence we don't need to subscribe to its effects once more.

After all of that, I'm thinking if it's a good thing at all. I'd say that it's not a good practice to provide effects on a component that'll be created in a way as you said. If you need that you have to provide effects once on the parent component level and make `app-todo` a `ui` component.

Reply:
Yes, ok.

Reply:
We can still define a global `Set` that stores a combination of the `constructor name + method` and warns the user on **dev** mode. What do u think?

Reply:
I think it could work if there'll be a thrown error. If the developer doesn't pay attention to the warning then we still will need to track the number of instances so we can remove from a set the last one when it's about to be destroyed.

Reply:
Ok, we can go with an error instead of warning

Reply:
I've just seen in NGRX docs next one: `Note: Registering an effects class multiple times, either by forRoot(), forFeature(), or [provideEffects](https://ngrx.io/api/effects/provideEffects)(), (for example in different lazy loaded features) will not cause the effects to run multiple times.`
Should we do the same for `provideEffects` and `forRoot`, `forFeature`? In this case, we don't need to throw an error for component effects since the class could have been already provided by `provideEffects`. We need one global `Set` which will contain `effectsClass.name + provideEffects.name(or useEffectsDirective.name)`.

Reply:
Yes, the same concept.

Reply:
But I'm unsure what you mean at the end of the comment. I think that constructor name + method is enough.

Reply:
There could be the case where the effects class was provided in `provideEffects` and `useEffectsDirective`. in this case, we can check whether `set` has `effectsClass.name + provideEffects.name` or not, and if it does we can safely skip providing in `useEffectsDirective`(and forget about unsubscribing on component destroy).
But... there could be another case where the effects class was provided with `useEffectsDirective` twice(none with `provideEffects`). In that one, we need to create the next token `effectsClass.name + useEffectsDirective.name` so that we understand that we need to unsubscribe from effects since the effects class was provided in the component injector.

Reply:
I'll push changes that I meant soon for you to understand.